synapse/logging/context.py

```diff
@@ -711,16 +735,19 @@ def nested_logging_context(suffix: str) -> LoggingContext:
     )
 
 
-def preserve_fn(f):
+def preserve_fn(f: F) -> F:
```

Comment:
We might want add a comment noting that the return type of `F` changes into a `Deferred[T]` if the original return type was not a `Deferred`.

I don't think we came to a consensus in #synapse-devs on whether it was okay to use slightly misleading type hints in this kind of situation.

Reply:
Ahh, I didn't spot that this also happens in `run_in_background`. Hmm, maybe this needs more thought?